### PR TITLE
refactor(core): typings and code cleanup

### PR DESCRIPTION
**Pull request by Wroud**
Part of #120

### CODE REVIEW

File: packages/core/src/createStore.ts [hunk 1]
```diff
@@ -1,26 +1,15 @@
...
+import { Atom, declareAtom, initAction, getState } from './declareAtom'
 
 type DispatchFunction = (action: Action<any>) => any
 type SubscribeFunction = {
   <T>(target: Atom<T>, listener: (state: T) => any): () => void
   (listener: DispatchFunction): () => void
 }
 type GetStateFunction = {
-  <T>(target: Atom<T>): T
+  <T>(target: Atom<T>): T | undefined
```
Comment (belozer) [inline packages/core/src/createStore.ts]: Why undefined?

Comment (belozer) [reply packages/core/src/createStore.ts]: GetState always returns state of atom in store or initial state

File: packages/core/src/createStore.ts [hunk 2]
```diff
@@ -68,19 +57,20 @@ export function createStore(
     }
   }
 
-  function _getState(target?: Atom<any>) {
+  function _getState(): State
+  function _getState<T>(target?: Atom<T>): State | T | undefined {
```
Comment (belozer) [inline packages/core/src/createStore.ts]: _getState not retruns `undefined`

File: packages/core/src/createStore.ts (hunk 1 above)
Comment (Wroud) [reply packages/core/src/createStore.ts]: it's related with `getState` function that can return undefined

File: packages/core/src/createStore.ts (hunk 2 above)
Comment (Wroud) [reply packages/core/src/createStore.ts]: if so we need change type of `getState` from `declareAtom.ts`

File: packages/core/src/declareAtom.ts [hunk 3]
```diff
@@ -74,8 +59,8 @@ export function declareAtom<State>(
     throwError(`Atom "${_id}". Initial state can't be undefined`)
 
   function reduce<T>(
-    dep: Unit<T>,
-    reducer: (state: State, payload: T) => State | undefined,
+    dep: Unit,
+    reducer: (state: TState, payload: T) => TState | undefined,
```
Comment (belozer) [inline packages/core/src/declareAtom.ts]: @artalar this is bug? `undefined` not alowed for state

File: packages/core/src/declareAtom.ts [hunk 4]
```diff
@@ -136,13 +121,10 @@ export function declareAtom<State>(
     }
   }
 
-  reduce(_initAction, (state = initialState) => state)
+  reduce(_initAction, (state = initialState as TState) => state)
```
Comment (belozer) [inline packages/core/src/declareAtom.ts]: Why not? `(state: TState = initialState) => state`

Comment (Wroud) [reply packages/core/src/declareAtom.ts]: here we need cast because initialState argument is union

File: packages/core/src/declareAtom.ts [hunk 5]
```diff
@@ -160,66 +140,64 @@ export function declareAtom<State>(
   return atom
 }
 
-export function getState<T>(state: State, atom: Atom<T>): T | undefined {
+export function getState<T>(state: State, atom: Atom<T>): T {
```
Comment (artalar) [inline packages/core/src/declareAtom.ts]: getState and store.getState is different functions. The getState must return optional type

File: packages/core/src/declareAtom.ts [hunk 6]
```diff
@@ -160,66 +140,64 @@ export function declareAtom<State>(
...
   return state[atom[TREE].id]
 }
 
-// @ts-ignore
-export declare function map<T, _T = unknown>(
+export function map<T, _T = unknown>(
```
Comment (artalar) [inline packages/core/src/declareAtom.ts]: Type inference for the map is broking
proof

File: packages/core/src/__tests__/index.ts [hunk 7]
```diff
@@ -360,13 +360,13 @@ describe('@reatom/core', () => {
...
     })
-    test('createStore with nullable atom', () => {
+    test('createStore with undefined atom', () => {
       const increment = declareAction()
       const countStatic = declareAtom(['countStatic'], 0, handle =>
         handle(increment, state => state + 1),
       )
 
-      const store = createStore(null, { countStatic: 10 })
+      const store = createStore(undefined, { countStatic: 10 })
```
Comment (belozer) [inline packages/core/src/__tests__/index.ts]: :confused: why not `null`? the code looks strange in this case

Comment (Wroud) [reply packages/core/src/__tests__/index.ts]: Function with null, cant have default value

Comment (belozer) [reply packages/core/src/__tests__/index.ts]: It's not like the usual code in all libraries.

Comment (Wroud) [reply packages/core/src/__tests__/index.ts]: You should not use null as undefined in javascript, null is value, but we used it as no value provided, we can't use default argument value for that

**Pull request merged**